src/lib/optimizer/dp_ccp.cpp

```diff
@@ -0,0 +1,182 @@
+#include "dp_ccp.hpp"
+
+#include <unordered_map>
+
+#include "cost_model/abstract_cost_estimator.hpp"
+#include "enumerate_ccp.hpp"
+#include "join_graph.hpp"
+#include "logical_query_plan/lqp_utils.hpp"
+#include "operators/operator_join_predicate.hpp"
+
+namespace opossum {
+
+DpCcp::DpCcp(const std::shared_ptr<AbstractCostEstimator>& cost_estimator) : _cost_estimator(cost_estimator) {}
+
+std::shared_ptr<AbstractLQPNode> DpCcp::operator()(const JoinGraph& join_graph) {
+  // No std::unordered_map, since hashing of JoinGraphVertexSet is not trivially possible because of hidden data
+  auto best_plan = std::map<JoinGraphVertexSet, std::shared_ptr<AbstractLQPNode>>{};
+
+  /**
+   * 1. Initialize single-vertex vertex_sets with the vertex nodes and their local predicates
+   */
+  for (size_t vertex_idx = 0; vertex_idx < join_graph.vertices.size(); ++vertex_idx) {
+    const auto vertex_predicates = join_graph.find_local_predicates(vertex_idx);
+    const auto vertex = join_graph.vertices[vertex_idx];
+
+    auto single_vertex_set = JoinGraphVertexSet{join_graph.vertices.size()};
+    single_vertex_set.set(vertex_idx);
+
+    best_plan[single_vertex_set] = _add_predicates_to_plan(vertex, vertex_predicates);
+  }
+
+  /**
+   * 2. Prepare EnumerateCcp: Transform the JoinGraph's vertex-to-vertex edges into index pairs
+   */
+  std::vector<std::pair<size_t, size_t>> enumerate_ccp_edges;
+  for (const auto& edge : join_graph.edges) {
+    // Don't include local predicate pseudo edges
+    if (edge.vertex_set.count() == 1) continue;
+    Assert(edge.vertex_set.count() == 2, "Can't place complex predicates yet");
+
+    const auto first_vertex_idx = edge.vertex_set.find_first();
+    const auto second_vertex_idx = edge.vertex_set.find_next(first_vertex_idx);
+
+    enumerate_ccp_edges.emplace_back(first_vertex_idx, second_vertex_idx);
+  }
+
+  /**
+   * 3. Actual DpCcp algorithm: Enumerate the CsgCmpPairs; build candidate plans; update best_plan
+   */
+  const auto csg_cmp_pairs = EnumerateCcp{join_graph.vertices.size(), enumerate_ccp_edges}();  // NOLINT
```

Comment:
CsgCmpPair is canon if you deal with join ordering algorithms, yes. I don't feel like explaining the concept in code is doing the reader a favour, since a more formal and correct explanation will always be in the paper.